python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -0,0 +1,265 @@
+                    *sources,
+                ]
+            except pydantic.ValidationError:
+                pass
+
+            metadata_by_key[key] = {
+                **metadata_by_key.get(key, {}),
+                **CodeReferencesMetadataSet(
+                    code_references=CodeReferencesMetadataValue(code_references=sources_for_asset)
+                ),
+            }
+
+    return assets_def.with_attributes(metadata_by_key=metadata_by_key)
+
+
+def convert_local_path_to_source_control_path(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    local_path: LocalFileCodeReference,
+) -> UrlCodeReference:
+    source_file_from_repo_root = os.path.relpath(
+        local_path.file_path, repository_root_absolute_path
+    )
+
+    return UrlCodeReference(
+        url=f"{base_source_control_url}/{source_file_from_repo_root}#L{local_path.line_number}",
+        label=local_path.label,
+    )
+
+
+def _convert_local_path_to_source_control_path_single_definition(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    assets_def: Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"],
+) -> Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"]:
+    from dagster._core.definitions.assets import AssetsDefinition
+
+    # SourceAsset doesn't have an op definition to point to - cacheable assets
+    # will be supported eventually but are a bit trickier
+    if not isinstance(assets_def, AssetsDefinition):
+        return assets_def
+
+    metadata_by_key = dict(assets_def.metadata_by_key) or {}
+
+    for key in assets_def.keys:
+        try:
+            existing_source_code_metadata = CodeReferencesMetadataSet.extract(
+                metadata_by_key.get(key, {})
+            )
+            sources_for_asset: List[Union[LocalFileCodeReference, UrlCodeReference]] = [
+                convert_local_path_to_source_control_path(
+                    base_source_control_url,
+                    repository_root_absolute_path,
+                    source,
+                )
+                if isinstance(source, LocalFileCodeReference)
+                else source
+                for source in existing_source_code_metadata.code_references.code_references
+            ]
+            metadata_by_key[key] = {
+                **metadata_by_key.get(key, {}),
+                **CodeReferencesMetadataSet(
+                    code_references=CodeReferencesMetadataValue(code_references=sources_for_asset)
+                ),
+            }
+        except pydantic.ValidationError:
+            pass
+
+    return assets_def.with_attributes(metadata_by_key=metadata_by_key)
+
+
+def _build_github_url(url: str, branch: str) -> str:
+    return f"{url}/tree/{branch}"
+
+
+def _build_gitlab_url(url: str, branch: str) -> str:
+    return f"{url}/-/tree/{branch}"
+
+
+@experimental
+def link_to_source_control(
```

Comment:
Could we potentially infer this information on behalf of the user by running a `git` command?

Reply:
Yes, or just walking the folder structure (since in some cases the user may not install git in the target image, in the cloud context at least). Stacked PR will have a layered utility fn that will have default behavior.